### ISSUE TITLE
sysext/SELinux: binaries are not labeled, only directories

### ISSUE THREAD

**Issue by thkukuk** [systemd/mkosi#3148]
### mkosi commit the issue has been seen with

main

### Used host distribution

openSUSE Tumbleweed

### Used target distribution

openSUSE Tumbleweed

### Linux kernel version used

6.11.5-1-default

### CPU architectures issue was seen on

x86_64

### Unexpected behaviour you saw

When building sysext images with SELinux enabled, the "base" image is fully labeled, all files and directories have a SELinux label.
But the sysext images itself are not. The directories contain a label (even the directories which are not part of the base image), but the files are unlabeled.
During the build process setfiles is called for all files, looks like the labels get lost later when copying the data. 
This is independent of if I use `sysext` as output format or `directory`.

```
# ls -alFZ /mnt/usr/bin/
total 15816
dr-xr-xr-x. 2 root root system_u:object_r:bin_t:s0      230 Okt 29 09:25 ./
drwxr-xr-x. 6 root root system_u:object_r:usr_t:s0       91 Okt 29 09:25 ../
-rwxr-xr-x  1 root root ?                             31904 Dez 14  2021 babeltrace*
-rwxr-xr-x  1 root root ?                             26744 Dez 14  2021 babeltrace-log*
-rwxr-xr-x  1 root root ?                              3503 Aug 29 07:55 gcore*
-rwxr-xr-x  1 root root ?                          16078368 Aug 29 07:55 gdb*
-rwxr-xr-x  1 root root ?                              4646 Aug 29 07:55 gdb-add-index*
-rwxr-xr-x  1 root root ?                              1063 Aug 29 07:55 gstack*
lrwxrwxrwx  1 root root ?                                 9 Sep  9 18:53 pydoc3 -> pydoc3.11*
-rwxr-xr-x  1 root root ?                                79 Sep  9 18:53 pydoc3.11*
lrwxrwxrwx  1 root root ?                                10 Sep  9 18:53 python3 -> python3.11*
-rwxr-xr-x  1 root root ?                             14448 Sep  9 18:53 python3.11*
#  ls -alFZ /mnt/usr/lib64/
total 6716
dr-xr-xr-x.  3 root root system_u:object_r:lib_t:s0     745 Okt 29 09:25 ./
drwxr-xr-x.  6 root root system_u:object_r:usr_t:s0      91 Okt 29 09:25 ../
lrwxrwxrwx   1 root root ?                               35 Dez 14  2021 libbabeltrace-ctf-metadata.so.1 -> libbabeltrace-ctf-metadata.so.1.0.0*
-rwxr-xr-x   1 root root ?                            14416 Dez 14  2021 libbabeltrace-ctf-metadata.so.1.0.0*
lrwxrwxrwx   1 root root ?                               26 Dez 14  2021 libbabeltrace-ctf.so.1 -> libbabeltrace-ctf.so.1.0.0*
-rwxr-xr-x   1 root root ?                           301752 Dez 14  2021 libbabeltrace-ctf.so.1.0.0*
lrwxrwxrwx   1 root root ?                               31 Dez 14  2021 libbabeltrace-ctf-text.so.1 -> libbabeltrace-ctf-text.so.1.0.0*
-rwxr-xr-x   1 root root ?                            43184 Dez 14  2021 libbabeltrace-ctf-text.so.1.0.0*
lrwxrwxrwx   1 root root ?                               28 Dez 14  2021 libbabeltrace-dummy.so.1 -> libbabeltrace-dummy.so.1.0.0*
-rwxr-xr-x   1 root root ?                            14408 Dez 14  2021 libbabeltrace-dummy.so.1.0.0*
lrwxrwxrwx   1 root root ?                               33 Dez 14  2021 libbabeltrace-lttng-live.so.1 -> libbabeltrace-lttng-live.so.1.0.0*
-rwxr-xr-x   1 root root ?                            43080 Dez 14  2021 libbabeltrace-lttng-live.so.1.0.0*
lrwxrwxrwx   1 root root ?                               22 Dez 14  2021 libbabeltrace.so.1 -> libbabeltrace.so.1.0.0*
-rwxr-xr-x   1 root root ?                            63528 Dez 14  2021 libbabeltrace.so.1.0.0*
-rwxr-xr-x   1 root root ?                            38960 Mai 21 17:45 libdebuginfod-0.191.so*
lrwxrwxrwx   1 root root ?                               22 Mai 21 17:45 libdebuginfod.so.1 -> libdebuginfod-0.191.so*
-rwxr-xr-x   1 root root ?                           199032 Sep  2  2021 libmpdec.so.2.5.1*
lrwxrwxrwx   1 root root ?                               17 Sep  2  2021 libmpdec.so.3 -> libmpdec.so.2.5.1*
-rwxr-xr-x   1 root root ?                          5306096 Sep  9 18:53 libpython3.11.so.1.0*
lrwxrwxrwx   1 root root ?                               28 Apr 26  2022 libsource-highlight.so.4 -> libsource-highlight.so.4.0.1*
-rwxr-xr-x   1 root root ?                           782296 Apr 26  2022 libsource-highlight.so.4.0.1*
drwxr-xr-x. 32 root root system_u:object_r:lib_t:s0    4435 Okt 29 09:25 python3.11/
```

### Used mkosi config

```ini
mkosi.conf:
[Distribution]
Distribution=opensuse
Release=tumbleweed

[Output]
OutputDirectory=mkosi.output
Format=none
ImageVersion=0.0

[Content]
SELinuxRelabel=true
Packages=
 MicroOS-release
 patterns-microos-basesystem
 patterns-microos-selinux
 gdb
 strace

mkosi.images/base/mkosi.conf:
[Output]
Format=directory
Output=base

[Content]
Packages=
 MicroOS-release
 patterns-microos-basesystem
 patterns-microos-selinux

mkosi.images/gdb/mkosi.conf:
[Config]
Dependencies=base

[Output]
Format=sysext
Overlay=yes
Output=gdb-%v.%a

[Validation]
Verity=no

[Content]
CleanPackageMetadata=yes
BaseTrees=%O/base
Packages=gdb
```

**Comment by DaanDeMeyer** [comment]
@thkukuk `SELinuxRelabel=` is not a universal setting, yet you only set it in mkosi.conf for the top level image. Because `SELinuxRelabel=` is a local setting, it needs to be configured individually for each image.

**Comment by thkukuk** [comment]
@DaanDeMeyer 
The setting in the top level image is only there to make sure everything for SELinux is available. If not, it will abort with this setting. So it cannot happen that by accident images without SELinux are build.
If everything is there, SELinux will be used for all images (`auto`). This works, as written, the files get labeled during build process and the directories are still labeled in the image.
Adding `SELinuxRelabel=` to all images does not make a different, as expected. For the base image, everything is labeled, for the sysext images, only directories are labeled, not files. I assume something with copying the files and overlayfs.

I will try to get debug output during the meetings...

**Comment by thkukuk** [comment]
mkosi running with --debug. But it looks like the really interesting part (overlayfs handling) has no debug output? Or is there an option to get even more verbose output?

[mkosi.log](https://github.com/user-attachments/files/17556946/mkosi.log)

**Comment by DaanDeMeyer** [comment]
So I can't reproduce this myself and offline @thkukuk mentioned that he was only able to reproduce it on one system and not another. So this is likely caused by an older kernel and not something we can work around in mkosi.

**Comment by thkukuk** [comment]
It's currently one working and more not working ;)
My current debugging results looks like as if it gets broken between `setfiles` and `mkosi.finalize`. So most likely either SELinux and the usual problems of labeling with a different policy than the global one or overlayfs. Since the labeling of the base image works, I still suspect it is overlayfs.
Since this is currently only a pet project by me, I don't have that much time to continue debugging.

**Comment by DaanDeMeyer** [comment]
@thkukuk Maybe you can list some information on the systems where it doesn't work and the system where it does work? I figure kernel versions will be the most interesting to look at.

**Comment by thkukuk** [comment]
Sorry for the late response, was busy with other tasks.
@DaanDeMeyer Comparing the systems was the first thing I did, all are up-to-date Tumbleweed installations, but with complete different setup. Comparing the setup is not that easy...
But I found it: if you run with security=apparmor, even if apparmor is not in use, setfiles does not work correct on overlayfs:
* labeling files and directories on btrfs works
* labeling directories on overlayfs works
* labeling files on overlayfs does **not** work.
This behavior is completely unexpected for me and doesn't make any sense. But sounds like an overlayfs bug.

Thanks for the help and patience.
